messaging/controller/controller.go

```diff
@@ -0,0 +1,225 @@
+package controller
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	gocommon "github.com/status-im/status-go/common"
+	"github.com/status-im/status-go/crypto"
+	"github.com/status-im/status-go/messaging/adapters"
+	"github.com/status-im/status-go/messaging/common"
+	"github.com/status-im/status-go/messaging/controller/processor"
+	"github.com/status-im/status-go/messaging/controller/sender"
+	"github.com/status-im/status-go/messaging/events"
+	"github.com/status-im/status-go/messaging/types"
+	"github.com/status-im/status-go/pkg/pubsub"
+)
+
+type Controller struct {
+	identity  *ecdsa.PrivateKey
+	stack     *common.MessagingStack
+	sender    *sender.Sender
+	processor *processor.Processor
+
+	messageConfirmationStorage common.MessageConfirmationPersistence
+	hashRatchetStorage         common.HashRatchetPersistence
+
+	publisher *pubsub.Publisher
+	logger    *zap.Logger
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+func NewController(
+	identity *ecdsa.PrivateKey,
+	stack *common.MessagingStack,
+	messageConfirmationStorage common.MessageConfirmationPersistence,
+	hashRatchetStorage common.HashRatchetPersistence,
+	publisher *pubsub.Publisher,
+	logger *zap.Logger,
+) *Controller {
+	return &Controller{
+		identity:                   identity,
+		stack:                      stack,
+		sender:                     sender.NewSender(identity, stack, logger),
+		processor:                  processor.NewProcessor(identity, stack, messageConfirmationStorage, hashRatchetStorage, logger),
+		messageConfirmationStorage: messageConfirmationStorage,
+		hashRatchetStorage:         hashRatchetStorage,
+		publisher:                  publisher,
+		logger:                     logger.Named("controller"),
+		quit:                       make(chan struct{}),
+	}
+}
+
+func (c *Controller) Start() error {
+	subscriptions, err := c.stack.Encryption.Start(c.identity)
+	if err != nil {
+		return err
+	}
+
+	// process stored shared secrets
+	err = c.processor.ProcessSharedSecrets(subscriptions.SharedSecrets)
+	if err != nil {
+		return err
+	}
+
+	err = c.StartReliability()
+	if err != nil {
+		return err
+	}
+
+	c.runSubscriptionsLoop()
+	c.runSegmentsCleanupLoop()
+	c.runHashRatchetCleanupLoop()
```

Comment:
Keep the functions flat internally, make them goroutines externally
```suggestion
    s.wg.Add(3)
	go c.subscriptionsLoop()
	go c.segmentsCleanupLoop()
	go c.hashRatchetCleanupLoop()
```